core/dbt/parser/schemas.py

```diff
@@ -914,6 +920,75 @@ def parse_patch(self, block: TargetBlock[NodeTarget], refs: ParserRef) -> None:
                 self.patch_node_config(node, patch)
 
             node.patch(patch)
+            self.validate_constraints(node)
+
+    def validate_constraints(self, patched_node):
+        error_messages = []
+        if (
+            patched_node.resource_type == "model"
+            and patched_node.config.constraints_enabled is True
+        ):
+            validators = [
+                self.constraints_schema_validator(patched_node),
+                self.constraints_materialization_validator(patched_node),
+                self.constraints_language_validator(patched_node),
+                self.constraints_data_type_validator(patched_node),
+            ]
+            error_messages = [validator for validator in validators if validator != "None"]
+
+        if error_messages:
+            original_file_path = patched_node.original_file_path
+            raise ParsingError(
+                f"Original File Path: ({original_file_path})\nConstraints must be defined in a `yml` schema configuration file like `schema.yml`.\nOnly the SQL table materialization is supported for constraints. \n`data_type` values must be defined for all columns and NOT be null or blank.{self.convert_errors_to_string(error_messages)}"
+            )
+
+    def convert_errors_to_string(self, error_messages: List[str]):
+        n = len(error_messages)
+        if not n:
+            return ""
+        if n == 1:
+            return error_messages[0]
+        error_messages_string = "".join(error_messages[:-1]) + f"{error_messages[-1]}"
+        return error_messages_string
+
+    def constraints_schema_validator(self, patched_node):
+        schema_error = False
+        if patched_node.columns == {}:
+            schema_error = True
+        schema_error_msg = "\n    Schema Error: `yml` configuration does NOT exist"
+        schema_error_msg_payload = f"{schema_error_msg if schema_error else None}"
+        return schema_error_msg_payload
+
+    def constraints_materialization_validator(self, patched_node):
+        materialization_error = {}
+        if patched_node.config.materialized != "table":
+            materialization_error = {"materialization": patched_node.config.materialized}
+        materialization_error_msg = f"\n    Materialization Error: {materialization_error}"
+        materialization_error_msg_payload = (
+            f"{materialization_error_msg if materialization_error else None}"
+        )
+        return materialization_error_msg_payload
+
+    def constraints_language_validator(self, patched_node):
+        language_error = {}
+        language = str(patched_node.language)
+        if language != "sql":
```

Comment:
_[For follow-up issues, out of scope for this PR]_

It should be _possible_ to support dbt-py models, by saving the Python model in a temp table, and then inserting it into a preexisting table (created per contract). The downside of that approach is, the full (potentially expensive) transformation has already taken place before the contract could be validated.

Reply:
I recommend the syntax be: `contract: true | false` because `contracted` invokes imagery of pregnancy(a beautiful human experience but dissonant related to data contracts) at first glance and is in past tense, which may confuse the reader. Whenever the term `data contracts` come up, it's a noun in plural or singular, not a verb. 

I'm on board with this example config, feels a lot more elegant compared to the first draft: 

```yml
columns:
  - name: price
    data_type: numeric
    constraints:
      - not null
      - check: price > 0
        name: positive_price  # Postgres supports naming constraints
```

I'm aligned with you on `get_columns_in_query` relying on the database cursor/API for more robustness and cheaper operations.

I'm aligned on python models as long as we document/warn the user of performance hits as a result. 

I'm aligned with you on "validators". My main goal is to provide maximally useful terminal logs that invoke clear and quick actions to fix. This makes iteration a delight rather than a chore. 

Thanks for such a thorough review and eye towards the future. I'm juiced up to finish these PRs!